File: userland/apps/asemics/rows.c
#include <stdlib.h>
#include <string.h>

#include "asemics.h"

void editorInsertRow(int pos, char *s, size_t len) {
    if (pos < 0 || pos > E.buffer.num_rows) return;

    E.buffer.rows = realloc(E.buffer.rows, sizeof(Row) * (E.buffer.num_rows + 1));
    memmove(&E.buffer.rows[pos + 1], &E.buffer.rows[pos], sizeof(Row) * (E.buffer.num_rows - pos));
    for (int j = pos + 1; j <= E.buffer.num_rows; j++)
        E.buffer.rows[j].idx++;

    E.buffer.rows[pos].idx = pos;

    E.buffer.rows[pos].size  = len;
    E.buffer.rows[pos].chars = malloc(len + 1);
    memcpy(E.buffer.rows[pos].chars, s, len);
    E.buffer.rows[pos].chars[len] = '\0';

    E.buffer.rows[pos].rsize           = 0;
    E.buffer.rows[pos].render          = NULL;
    E.buffer.rows[pos].hl              = NULL;
    E.buffer.rows[pos].hl_open_comment = 0;
    editorUpdateRow(&E.buffer.rows[pos]);

    E.buffer.num_rows++;
}

void editorInsertNewLine(void) {
    if (E.cursor.x == 0) {
        editorInsertRow(E.cursor.y, "", 0);
    } else {
        Row *er = &E.buffer.rows[E.cursor.y];
        editorInsertRow(E.cursor.y + 1, &er->chars[E.cursor.x], er->size - E.cursor.x);
        er                  = &E.buffer.rows[E.cursor.y];
        er->size            = E.cursor.x;
        er->chars[er->size] = '\0';
        editorUpdateRow(er);
    }
    E.cursor.y++;
    E.cursor.x = 0;
}

void editorUpdateRow(Row *er) {
    int tabs = 0;
    int j;
    for (j = 0; j < er->size; j++)
        if (er->chars[j] == '\t') tabs++;

    free(er->render);
    er->render = malloc(er->size + tabs * (TAB_STOP - 1) + 1);

    int idx = 0;
    for (j = 0; j < er->size; j++) {
        if (er->chars[j] == '\t') {
            er->render[idx++] = ' ';
            while (idx % TAB_STOP != 0)
                er->render[idx++] = ' ';
        } else {
            er->render[idx++] = er->chars[j];
        }
    }

    er->render[idx] = '\0';
    er->rsize       = idx;

    editorUpdateSyntax(er);
}

int editorRowXtoRx(Row *er, int x) {
    int rx = 0;
    int j  = 0;

    for (j = 0; j < x; j++) {
        if (er->chars[j] == '\t') rx += (TAB_STOP - 1) - (rx % TAB_STOP);
        rx++;
    }
    return rx;
}

int editorRowRxToX(Row *row, int rx) {
    int cur_rx = 0;
    int cx;
    for (cx = 0; cx < row->size; cx++) {
        if (row->chars[cx] == '\t') cur_rx += (TAB_STOP - 1) - (cur_rx % TAB_STOP);
        cur_rx++;

        if (cur_rx > rx) return cx;
    }
    return cx;
}

void editorRowInsertChar(Row *er, int pos, int c) {
    if (pos < 0 || pos > er->size) pos = er->size;
    er->chars = realloc(er->chars, er->size + 2);
    memmove(&er->chars[pos + 1], &er->chars[pos], er->size - pos + 1);
    er->size++;
    er->chars[pos] = c;
    editorUpdateRow(er);
}

void editorInsertChar(int c) {
    if (E.cursor.y == E.buffer.num_rows) { editorInsertRow(E.buffer.num_rows, "", 0); }
    editorRowInsertChar(&E.buffer.rows[E.cursor.y], E.cursor.x, c);
    E.cursor.x++;
    E.buffer.dirty = true;
}

void editorFreeRow(Row *er) {
    free(er->render);
    free(er->chars);
    free(er->hl);
}

void editorDelRow(int pos) {
    if (pos < 0 || pos >= E.buffer.num_rows) return;
    editorFreeRow(&E.buffer.rows[pos]);
    memmove(&E.buffer.rows[pos], &E.buffer.rows[pos + 1],
            sizeof(Row) * (E.buffer.num_rows - pos - 1));
    for (int j = pos; j < E.buffer.num_rows - 1; j++)
        E.buffer.rows[j].idx--;
    E.buffer.num_rows--;
    E.buffer.dirty = true;
}

void editorRowDelChar(Row *er, int pos) {
    if (pos < 0 || pos >= er->size) return;
    memmove(&er->chars[pos], &er->chars[pos + 1], er->size - pos);
    er->size--;
    editorUpdateRow(er);
    E.buffer.dirty = true;
}

void editorDelChar(void) {
    if (E.cursor.y == E.buffer.num_rows) return;
    if (E.cursor.x == 0 && E.cursor.y == 0) return;

    Row *er = &E.buffer.rows[E.cursor.y];
    if (E.cursor.x > 0) {
        editorRowDelChar(er, E.cursor.x - 1);
        E.cursor.x--;
    } else {
        E.cursor.x = E.buffer.rows[E.cursor.y - 1].size;
        editorRowAppendString(&E.buffer.rows[E.cursor.y - 1], er->chars, er->size);
        editorDelRow(E.cursor.y);
        E.cursor.y--;
    }
}

void editorRowAppendString(Row *er, char *s, size_t len) {
    er->chars = realloc(er->chars, er->size + len + 1);
    memcpy(&er->chars[er->size], s, len);
    er->size += len;
    er->chars[er->size] = '\0';
    editorUpdateRow(er);
    E.buffer.dirty = true;
}

File: userland/apps/asemics-rs/src/rows.rs
use crate::consts::TAB_STOP;
use crate::editor::Editor;

#[derive(Clone)]
pub struct Row {
    pub idx: i32,             /* this row's position in the buffer */
    pub chars: Vec<u8>,       /* raw line text as stored in the file */
    pub render: Vec<u8>,      /* line as drawn on screen (tabs expanded etc.) */
    pub hl: Vec<u8>,          /* per-cell highlight codes, same length as `render` */
    pub hl_open_comment: bool, /* true if this row ends inside an unclosed multi-line comment */
}

impl Row {
    /* the C code tracks size/rsize alongside the buffers; here they are always
     * the buffer lengths */
    pub fn size(&self) -> i32 {
        self.chars.len() as i32
    }

    pub fn rsize(&self) -> i32 {
        self.render.len() as i32
    }

    pub fn x_to_rx(&self, x: i32) -> i32 {
        let mut rx = 0;
        for j in 0..x {
            if (j as usize) < self.chars.len() && self.chars[j as usize] == b'\t' {
                rx += (TAB_STOP - 1) - (rx % TAB_STOP);
            }
            rx += 1;
        }
        rx
    }

    pub fn rx_to_x(&self, rx: i32) -> i32 {
        let mut cur_rx = 0;
        let mut cx = 0;
        while cx < self.size() {
            if self.chars[cx as usize] == b'\t' {
                cur_rx += (TAB_STOP - 1) - (cur_rx % TAB_STOP);
            }
            cur_rx += 1;

            if cur_rx > rx {
                return cx;
            }
            cx += 1;
        }
        cx
    }
}

impl Editor {
    pub fn insert_row(&mut self, pos: i32, s: &[u8]) {
        if pos < 0 || pos > self.buffer.num_rows {
            return;
        }

        let row = Row {
            idx: pos,
            chars: s.to_vec(),
            render: Vec::new(),
            hl: Vec::new(),
            hl_open_comment: false,
        };
        self.buffer.rows.insert(pos as usize, row);
        for j in (pos + 1) as usize..self.buffer.rows.len() {
            self.buffer.rows[j].idx += 1;
        }

        self.update_row(pos);

        self.buffer.num_rows += 1;
    }

    pub fn insert_new_line(&mut self) {
        if self.cursor.x == 0 {
            self.insert_row(self.cursor.y, b"");
        } else {
            let y = self.cursor.y as usize;
            let x = self.cursor.x as usize;
            let rest = self.buffer.rows[y].chars[x..].to_vec();
            self.insert_row(self.cursor.y + 1, &rest);
            self.buffer.rows[y].chars.truncate(x);
            self.update_row(self.cursor.y);
        }
        self.cursor.y += 1;
        self.cursor.x = 0;
    }

    pub fn update_row(&mut self, idx: i32) {
        let row = &mut self.buffer.rows[idx as usize];

        let mut render = Vec::with_capacity(row.chars.len());
        for &c in &row.chars {
            if c == b'\t' {
                render.push(b' ');
                while render.len() % TAB_STOP as usize != 0 {
                    render.push(b' ');
                }
            } else {
                render.push(c);
            }
        }
        row.render = render;

        self.update_syntax(idx);
    }

    pub fn row_insert_char(&mut self, idx: i32, pos: i32, c: i32) {
        let row = &mut self.buffer.rows[idx as usize];
        let mut pos = pos;
        if pos < 0 || pos > row.size() {
            pos = row.size();
        }
        row.chars.insert(pos as usize, c as u8);
        self.update_row(idx);
    }

    pub fn insert_char(&mut self, c: i32) {
        if self.cursor.y == self.buffer.num_rows {
            self.insert_row(self.buffer.num_rows, b"");
        }
        self.row_insert_char(self.cursor.y, self.cursor.x, c);
        self.cursor.x += 1;
        self.buffer.dirty = true;
    }

    pub fn del_row(&mut self, pos: i32) {
        if pos < 0 || pos >= self.buffer.num_rows {
            return;
        }
        self.buffer.rows.remove(pos as usize);
        for j in pos as usize..self.buffer.rows.len() {
            self.buffer.rows[j].idx -= 1;
        }
        self.buffer.num_rows -= 1;
        self.buffer.dirty = true;
    }

    pub fn row_del_char(&mut self, idx: i32, pos: i32) {
        let row = &mut self.buffer.rows[idx as usize];
        if pos < 0 || pos >= row.size() {
            return;
        }
        row.chars.remove(pos as usize);
        self.update_row(idx);
        self.buffer.dirty = true;
    }

    pub fn del_char(&mut self) {
        if self.cursor.y == self.buffer.num_rows {
            return;
        }
        if self.cursor.x == 0 && self.cursor.y == 0 {
            return;
        }

        if self.cursor.x > 0 {
            self.row_del_char(self.cursor.y, self.cursor.x - 1);
            self.cursor.x -= 1;
        } else {
            let y = self.cursor.y as usize;
            self.cursor.x = self.buffer.rows[y - 1].size();
            let s = self.buffer.rows[y].chars.clone();
            self.row_append_string(self.cursor.y - 1, &s);
            self.del_row(self.cursor.y);
            self.cursor.y -= 1;
        }
    }

    pub fn row_append_string(&mut self, idx: i32, s: &[u8]) {
        self.buffer.rows[idx as usize].chars.extend_from_slice(s);
        self.update_row(idx);
        self.buffer.dirty = true;
    }
}
